internal/runner/runner.go

```diff
@@ -60,68 +57,61 @@ func (r *runner) Run(ctx context.Context) error {
 		return err
 	}
 
-	namespaceList, err := r.cs.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	// get list of pods from all namespaces
+	podList, err := r.cs.CoreV1().Pods("").List(cancelCtx, metav1.ListOptions{})
 	if err != nil {
 		return err
 	}
 
-	for _, ns := range namespaceList.Items {
-		podList, err := r.cs.CoreV1().Pods(ns.Name).List(ctx, metav1.ListOptions{})
-		if err != nil {
-			return err
-		}
+	for _, pod := range podList.Items {
+		go r.streamPodLogs(cancelCtx, r.cs, pod)
 
-		for _, pod := range podList.Items {
-
-			for _, container := range pod.Spec.Containers {
-				req := r.cs.CoreV1().Pods(ns.Name).GetLogs(pod.Name, &corev1.PodLogOptions{
-					Timestamps: true,
-					Container:  container.Name,
-				})
-
-				podLogs, err := req.Stream(ctx)
-				if err != nil {
-					return err
-				}
-				defer podLogs.Close()
-
-				buf := new(bytes.Buffer)
-
-				_, err = io.Copy(buf, podLogs)
-				if err != nil {
-					return err
-				}
-
-				labels := make(map[string]string)
-				labels["namespace"] = ns.Name
-				labels["pod_name"] = pod.Name
-				labels["container_name"] = container.Name
-
-				err = r.loadLogsToLoki(buf, parser.NewContainerParser(), labels)
-				if err != nil {
-					return err
-				}
-			}
-		}
 	}
 
 	return nil
 }
 
-func (r *runner) loadLogsToLoki(rawLogs *bytes.Buffer, logParser parser.Parser, labels map[string]string) error {
+func (r *runner) loadLogsToLoki(logLine string, logParser parser.Parser, labels map[string]string) error {
+	tm, labelset, err := logParser.Parse(logLine, labels)
+	if err != nil {
+		r.logger.Log("Skipping log due to invalid parse", "Error", err.Error())
+		return err
+	}
+	r.lokiClient.PostLog(logLine, tm, labelset)
 
-	scanner := bufio.NewScanner(rawLogs)
-	scanner.Split(bufio.ScanLines)
+	return nil
+}
 
-	for scanner.Scan() {
-		log_line := scanner.Text()
-		tm, labels, err := logParser.Parse(log_line, labels)
+// streamPodLogs will stream the pod logs and load the logs to loki with relevant
+// labels, loglines and timestamp
+func (r *runner) streamPodLogs(cancelCtx context.Context, cs kubernetes.Interface, pod corev1.Pod) error {
+	for _, container := range pod.Spec.Containers {
+		podLogOptions := &corev1.PodLogOptions{
+			Follow:     true,
+			Timestamps: true,
+		}
+
+		req := cs.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, podLogOptions)
+		stream, err := req.Stream(cancelCtx)
 		if err != nil {
-			r.logger.Log("Skipping log due to invalid parse", "Error", err.Error())
-			continue
+			return err
 		}
-		r.lokiClient.PostLog(log_line, tm, labels)
-	}
 
+		reader := bufio.NewScanner(stream)
+		reader.Split(bufio.ScanLines)
+		defer stream.Close()
+
+		for reader.Scan() {
```

Comment:
because the `Follow` attribute in the `podLogOptions` has been set to `true`, this `reader.Scan()` will always be true and reads continuously (and hence can be used as infinite loop and reader too)